wumpus_world.py

```diff
@@ -118,21 +117,154 @@ def _transform_kb(self):
                     facts.add(sentence)
         
         return facts, implications, bidirectionals
+    
+    def inference_by_resolution(self, query):
+        """
+        This will be the main method that you will call to
+        perform inference by resolution.
+        """
+        # CNF The kb and add the negated query to it
+        kb_with_negated_query = [self.sentence_to_cnf(i) for i in self.kb]
+        kb_with_negated_query.append(('NOT', self.sentence_to_cnf(query)))
+        
+        while True:
+            # Comparing each clause with every other clause in the kb
+            for i in range(len(kb_with_negated_query) - 1):
+                for j in range(i + 1, len(kb_with_negated_query)):
+                    resolved = self.resolve(kb_with_negated_query[i], kb_with_negated_query[j])
+
+                    # Check if this is the last possible pair of clauses, if it is we return True when the kb contradicts itself and False when it doesn't
+                    if set(resolved).issubset(set(kb_with_negated_query)) and i + 2 == len(kb_with_negated_query) and j + 1 == len(kb_with_negated_query):
+                        if query in kb_with_negated_query and ('NOT', query) in kb_with_negated_query:
+                            return True
+                        else:
+                            return False
+                        
+                    # If this is not the last possible pair of clauses, simply add the new clause to the kb if it isn't already
+                    else:
+                        for clause in resolved:
+                            if clause not in kb_with_negated_query:
+                                kb_with_negated_query.append(clause)
+
```

Comment:
I may be wrong, but I think there's a _potential_ issue with this implementation. When the for loops execute for the first time, the `len(kb_with_negated_query)` is calculated for once. That value will not change even after you add new resolved clauses to `kb_with_negated_query`.  Therefore, this implementation doesn't really resolve every pairs of clauses possible.